google/cloud/internal/backoff_policy.h

```diff
@@ -0,0 +1,151 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_BACKOFF_POLICY_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_BACKOFF_POLICY_H_
+
+#include "google/cloud/internal/random.h"
+#include "google/cloud/internal/throw_delegate.h"
+#include <chrono>
+#include <memory>
+
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+/**
+ * Define the interface for backoff policies.
+ *
+ * The client libraries need to hide partial and temporary failures from the
+ * application. Exponential backoff is generally considered a best practice when
+ * retrying operations. However, the details of how exponetial backoff is
+ * implemented and tuned varies widely. We need to give the users enough
+ * flexibility, and also provide sensible default implementations.
+ *
+ * The client library receives an object of this type, and clones a new instance
+ * for each operation. That is, the application provides the library with a
+ * [Prototype](https://en.wikipedia.org/wiki/Prototype_pattern) of the policy
+ * that will be applied to each operation.
+ *
+ * [Truncated Exponential
+ * Backoff](https://cloud.google.com/storage/docs/exponential-backoff) in the
+ * Google Cloud Storage documentation.
+ *
+ */
+class BackoffPolicy {
+ public:
+  virtual ~BackoffPolicy() = default;
+
+  /**
+   * Return a new copy of this object.
+   *
+   * Typically implemented as
+   * @code
+   *   return std::unique_ptr<BackoffPolicy>(new Foo(*this));
+   * @endcode
+   */
+  virtual std::unique_ptr<BackoffPolicy> clone() const = 0;
+
+  /**
+   * Return the delay after an RPC operation has completed.
```

Comment:
Suggested docstring change --> "Return the delay (in ms) to wait before the next retry attempt."

Reply:
Generally I like this, with a few nits:

- The 'Return' thing is already put [there](https://googlecloudplatform.github.io/google-cloud-cpp/0.1.0/bigtable/classbigtable_1_1v0_1_1ExponentialBackoffPolicy.html#a48e212f5e83c54fc19630afb63d7ef14) by Doxygen
- The return type is called `std::chrono::milliseconds` is it worth it to repeat the units?  I will change it if you think it is.